go/libraries/doltcore/sqle/binlogreplication/binlog_replica_applier.go

```diff
@@ -0,0 +1,997 @@
+			"database":  tableMap.Database,
+			"flags":     convertToHexString(tableMap.Flags),
+			"metadata":  tableMap.Metadata,
+			"types":     tableMap.Types,
+		}).Debug("Received binlog event: TableMap")
+
+		if tableId == 0xFFFFFF {
+			// Table ID 0xFFFFFF is a special value that indicates table maps can be freed.
+			logger.Infof("binlog protocol message: table ID '0xFFFFFF'; clearing table maps")
+			a.tableMapsById = make(map[uint64]*mysql.TableMap)
+		} else {
+			flags := tableMap.Flags
+			if flags&rowFlag_endOfStatement == rowFlag_endOfStatement {
+				// nothing to be done for end of statement; just clear the flag
+				flags = flags &^ rowFlag_endOfStatement
+			}
+			if flags&rowFlag_noForeignKeyChecks == rowFlag_noForeignKeyChecks {
+				flags = flags &^ rowFlag_noForeignKeyChecks
+			}
+			if flags != 0 {
+				msg := fmt.Sprintf("unsupported binlog protocol message: TableMap event with unsupported flags '%x'", flags)
+				logger.Errorf(msg)
+				DoltBinlogReplicaController.setSqlError(mysql.ERUnknownError, msg)
+			}
+			a.tableMapsById[tableId] = tableMap
+		}
+
+	case event.IsDeleteRows():
+		// A ROWS_EVENT is written for row based replication if data is inserted, deleted or updated.
+		// For more details, see: https://mariadb.com/kb/en/rows_event_v1v2-rows_compressed_event_v1/
+		logger.Debug("Received binlog event: DeleteRows")
+		tableId := event.TableID(a.format)
+		tableMap, ok := a.tableMapsById[tableId]
+		if !ok {
+			return fmt.Errorf("unable to find replication metadata for table ID: %d", tableId)
+		}
+
+		if a.filters.isTableFilteredOut(tableMap) {
+			return nil
+		}
+		a.modifiedDatabases[tableMap.Database] = struct{}{}
+
+		rows, err := event.Rows(a.format, tableMap)
+		if err != nil {
+			return err
+		}
+
+		flags := rows.Flags
+		if flags&rowFlag_endOfStatement == rowFlag_endOfStatement {
+			// nothing to be done for end of statement; just clear the flag and move on
+			flags = flags &^ rowFlag_endOfStatement
+		}
+		if flags&rowFlag_noForeignKeyChecks == rowFlag_noForeignKeyChecks {
+			flags = flags &^ rowFlag_noForeignKeyChecks
+		}
+		if flags != 0 {
+			msg := fmt.Sprintf("unsupported binlog protocol message: DeleteRows event with unsupported flags '%x'", flags)
+			logger.Errorf(msg)
+			DoltBinlogReplicaController.setSqlError(mysql.ERUnknownError, msg)
+		}
+		schema, err := getTableSchema(ctx, engine, tableMap.Name, tableMap.Database)
+		if err != nil {
+			return err
+		}
+
+		logger.Debugf(" - Deleted Rows (table: %s)", tableMap.Name)
+		for _, row := range rows.Rows {
+			deletedRow, err := parseRow(ctx, tableMap, schema, rows.IdentifyColumns, row.NullIdentifyColumns, row.Identify)
+			if err != nil {
+				return err
+			}
+			logger.Debugf("     - Identify: %v ", sql.FormatRow(deletedRow))
+
+			foreignKeyChecksDisabled := tableMap.Flags&rowFlag_noForeignKeyChecks > 0
+			writeSession, tableWriter, err := getTableWriter(ctx, engine, tableMap.Name, tableMap.Database, foreignKeyChecksDisabled)
```

Comment:
Some more details in the next comment... but essentially, I combined the update/insert/delete logic to reduce duplication and I batched the WriteSession per-table instead of flushing for each row. A step in the right direction, but I'd still like to take another pass and come back and see if I can debug the session-managed WriteSession issues I hit yesterday.